include/bluetooth/audio/hap.h

```diff
@@ -0,0 +1,275 @@
+ * The profile instance is removed when the bond is removed or when @ref
+ * bt_hap_delete is called.
+ * 
+ * @param conn The connection to initialize the profile for.
+ *
+ * @return 0 in case of success or negative value in case of error.
+ */
+int bt_hap_new(struct bt_conn *conn);
+
+/**
+ * @brief Delete profile instance
+ *
+ * This will remove stored profile instance of remote Hearing Aid Service.
+ *
+ * @param hap Profile instance object.
+ *
+ * @return 0 in case of success or negative value in case of error.
+ */
+int bt_hap_delete(struct bt_hap *hap);
+
+/**
+ * @brief Iterate through all existing profile instances.
+ *
+ * Iterate profile instances of remote Hearing Access Services.
+ *
+ * @param func Function to call for each profile instance.
+ * @param user_data Data to pass to the callback function.
+ */
+void bt_hap_foreach(bool (*func)(const struct bt_hap *hap, void *user_data),
+		    void *user_data);
+
+/**
+ * @brief Get profile instance to local Hearing Access Service.
+ *
+ * Get the profile instance connection to local Hearing Access Service.
+ *
+ * @return Profile instance object or NULL if there is no service.
+ */
+struct bt_hap *bt_hap_local_get(void);
+
+/**
+ * @brief Get the Bluetooth connection object of the profile instance.
+ *
+ * Get the Bluetooth connection object of a HAP instance.
+ *
+ * @param hap Profile instance object.
+ *
+ * @return Bluetooth connection object or NULL.
+ */
+struct bt_conn *bt_hap_conn_get(const struct bt_hap *hap);
+
+/** 
+ * @typedef bt_hap_complete_func_t
+ * @brief Operation complete callback.
+ *
+ * @param hap Profile instance object.
+ * @param err Error code.
+ * @param user_data User data passed to callback.
+ */
+typedef void (*bt_hap_complete_func_t) (struct bt_hap *hap, int err,
+					void *user_data);
+
+/**
+ * @brief Write Preset Name.
+ *
+ * Write the Name of a writable preset record identified by the Index.
+ *
+ * @param hap Profile instance object.
+ * @param index Index of preset record that name to be written.
+ * @param data Name to be written.
+ * @param length Name length.
+ * @param func Operation complete callback.
+ * @param user_data User data to be passed back to callback.
+ *
+ * @return 0 in case of success or negative value in case of error.
+ */
+int bt_hap_preset_write_name(struct bt_hap *hap, uint8_t index,
+			     const void *data, uint8_t length,
+			     bt_hap_complete_func_t func, void *user_data);
+
+/**
+ * @brief Set Active Preset.
+ *
+ * Set the preset record identified by the Index field as the active preset.
+ *
+ * @param hap Profile instance object.
+ * @param index Preset record index to be set to active.
+ * @param func Operation complete callback.
+ * @param user_data User data to be passed back to callback.
+ *
+ * @return 0 in case of success or negative value in case of error.
+ */
+int bt_hap_preset_set_active(struct bt_hap *hap, uint8_t index,
+		      	     bt_hap_complete_func_t func, void *user_data);
```

Comment:
It atypical to provide a specific callback for each function in Zephyr BT IIRC. I think GATT may be the only feature that uses that, where the rest uses "global" callbacks. 

We could consider moving the `bt_hap_complete_func` to the `bt_hap_cb` structure, and simply use that. It will be less flexible if we just keep a single callback for all the procedures, so alternatively we can define a callback for each procedure in those (similar to what has been done in e.g. `bt_vcs_cb`

Reply:
The reason why there is specific callback for each function is because I wanted that only `bt_hap_preset_set_active` caller will get the status of the operation (ATT Write Response to be more precise). The API does not restrict the number of users to be 1 thus BT_HAP_CB_DEFINE allows to register a number of callbacks.

Reply:
> It would be kind'a odd, if the bt_hap_complete_func would be called to the user that did not called bt_hap_preset_set_active

That's arguably true, but that is what we do for the `struct bt_conn_cb` callbacks, which arguably is what closest resemble these callbacks as you also define the `BT_HAP_CB_DEFINE`. 

What we do for the remaining (scan, adv, l2cap, vcs, etc.), is that we pass a callback structure when we initialize the struct. For example, instead of a global callback, you could provide the `bt_hap_cb` to `bt_hap_new`, so that each profile instance has it own set of callbacks/ops.